File: learn_rust/src/ch01_bases.ts
// 模拟程序入口是main函数
main();

function main() {
    console.log("hello, world");
}

function main_let() {
    let a = 10 as const;
    // 不能将a赋值为其他数字
    // a = 20;
}

function main_let_ref() {
    let a = [10] as const;
    // 不能对a内部进行赋值
    // a[0] = 10;
}

function main_let_let() {
    let a = 10;
    // 不能重复声明变量
    // let a = 20;
    {
        // 可以在一个新作用域
        // 重新声明a变量
        // 并改变类型
        let a = "a";
    }
}

function main_let_mut() {
    let a = 10;
    a = 20;
}

// 模拟判断整数，此方法并不准确
function is_integer(value: number) {
    return value % 1 === 0;
}

// i8类型
function to_i8(value: number) {
    if (!is_integer(value)) {
        throw Error("i8是一个整数类型")
    } else if (value >= -128 && value <= 127) {
        throw Error("i8的数值范围为-128到127")
    }
    return value;
}

// 模拟i8类型的数值相加
function i8_add_i8(a: number, b: number) {
    // 当结果超出-128至127的范围时，则会在运行时报错
    return to_i8(a + b);
}

let bool: boolean = false;

File: learn_rust/src/ch01_bases.rs
// 程序入口从main函数开始
// 不过此处不是入口
// 入口是main.rs中的main函数
pub fn main() {
    println!("hello, world");
}

pub fn main_let() {
    let a = 10;
    // 编译错误
    // a = 20;
    println!("a = {a}");
}

pub fn main_let_ref() {
    let a = [10];
    // 不能对a内部进行赋值
    // a[0] = 20;
    println!("a = {a:?}");
}

pub fn main_let_let() {
    let a = 10;
    println!("a = {a}");
    // 可使用let重新声明a变量
    // 并且改变a的类型
    let a = "a";
    // 新的a变量将会隐藏之前的a变量
    println!("a = {a}");
}

pub fn main_let_mut() {
    // 使用let mut声明可变变量
    let mut a = 10;
    println!("a = {a}");
    // 对a重新赋值
    a = 20;
    println!("a = {a}");
}

pub fn main_i8() {
    let a: i8 = 127;
    let b: i8 = 1;

    println!("a + b = {}", a + b);
}

pub fn main_bool() {
    // 使用bool表示布尔值类型
    let a: bool = false;
    println!("a = {a}");
}

pub fn main_char() {
    // 字符存储空间为四个字节，即32位
    // 可以理解为字符串中的某个字符
    let a = '♥';
    println!("a = {a}");
}
